iocore/nexthop/NextHopHost.h

```diff
@@ -0,0 +1,201 @@
+ * Plugins should define property blocks.
+ * @see PropBlockDeclare
+ */
+struct HostAddr : public PropertyBlock<HostAddr> {
+  const IpEndpoint m_addr;      ///< ip and port
+  std::atomic_uint m_eol;       ///< "end of life" time, set by DNS time to live
+  std::atomic_bool m_available; ///< true when this IP is available for use
+
+  HostAddr(const IpEndpoint &ip_addr, HostNamePtr host) : m_addr(ip_addr) { HostLookupByAddr.put(ip_addr, host); }
+
+  // Note, this uses PropertyBlock::new and delete to manage.
+
+private:
+  HostAddr();
+};
+
+//////////////////////////////////////////////
+
+/// Defines a host (name), group of IPs, and block of associated properties.
+/** Only variables need by TSCore are in the structure.
+ * Plugins can use property blocks to store host associated data.
+ * @see PropBlockDeclare
+ */
+class HostName : public PropertyBlock<HostName>
+{
+public:
+  // to enforce all references be shared_ptr
+  static HostNamePtr
+  alloc(ts::string_view hostname)
+  {
+    HostID id = std::hash<ts::string_view>()(hostname);
+
+    // get a lock incase we are allocating hosts in parallel
+    LockGuard part_lock(s_lock_pool.getMutex(s_lock_pool.getIndex(id)));
+
+    // return existing host by name, to prevent duplicates
+    HostNamePtr existing_host = getHost(id);
+    if (existing_host) {
+      // increment the count of systems that require this to exist.
+      existing_host->m_strong_ref_count++;
+      return existing_host;
+    }
+
+    HostNamePtr host(new HostName(hostname)); ///< call PropertyBlock::operator new
+    host->propBlockInit();
+
+    HostLookupByNameHash.put(id, host);
+    return host;
+  }
+
+  void
+  free()
+  {
+    ink_assert(m_strong_ref_count > 0); // we are calling 'free' more than 'alloc'
+    ink_assert(hasWriteAccess(this));   // get the lock first, getMutex().lock()
+
+    // decrement the count of systems that require this to exist.
+    m_strong_ref_count--;
+    if (m_strong_ref_count > 0) {
+      return;
+    }
+
+    // start cleaning up this reference
+    HostID id = std::hash<std::string>()(m_name);
+    HostLookupByNameHash.erase(id);
+    propBlockDestroy();
+    getMutex().unlock();
+
+    // it will delete after shared_ptr's are cleared,
+    // most of which should have happend in the above propBlockDestroy()
+    // or when current
+  }
+
+  ~HostName() { ink_assert(m_strong_ref_count == 0); }
+
+#if DEBUG
+  static bool
+  hasReadAccess(const HostName *host)
+  {
+    auto m = host->getMutex();
+    return m.has_lock_shared() || m.has_lock();
+  }
+  static bool
+  hasWriteAccess(const HostName *host)
+  {
+    return host->getMutex().has_lock();
+  }
+#endif
+  /// ----------------------------
+  /// Thread Safe Methods
+  ///
+
+  /// thread management
+  Mutex &
+  getMutex() const
+  {
+    return s_lock_pool.getMutex(m_lock_idx);
+  }
+
+  /// get the FQDN that defines this host.
+  const ts::string_view
+  getName() const
+  {
+    return ts::string_view(m_name);
```

Comment:
Did you try `return m_name` or `return {m_name}`?

Reply:
no. don't they all compile to the same thing?